image/postgres/scripts/init-entrypoint.sh

```diff
@@ -5,4 +5,50 @@ set -Eeo pipefail
 # Initialize DB if it does not exist
 if [ ! -s "$PGDATA/PG_VERSION" ]; then
   initdb --auth-host=scram-sha-256 --auth-local=scram-sha-256 --pwfile /run/secrets/stackrox.io/secrets/password --data-checksums
+else
+    # Verify if we need to perform major version upgrade
+    PG_BINARY_VERSION=$(postgres -V |\
+        sed 's/postgres (PostgreSQL) \([0-9]*\).\([0-9]*\).*/\1/')
+
+    if [ $(cat "${PGDATA}/PG_VERSION") -lt "$PG_BINARY_VERSION" ]; then
+        # Data version is less than binaries, upgrade
+
+        # Do a backup first, ideally move it to a separate volume. Since the
+        # database is stopped we may as well simple take a filesystem backup.
+        # Alternatives would be pg_dump or pg_basebackup, both require running
+        # database cluster.
+        echo "Backup..."
+        export BACKUP_DIR="/tmp/backups/$(date +%s)"
+        mkdir -p $BACKUP_DIR
+        tar -cf $BACKUP_DIR/backup.tar -C $PGDATA .
+        sync $BACKUP_DIR/backup.tar
+
+        echo "Verify backup..."
+        export BACKUP_VERIFY_PGDATA="/tmp/backup-test"
+        export OLD_BINARIES="/usr/lib64/pgsql/postgresql-13/bin/"
+        mkdir -p $BACKUP_VERIFY_PGDATA
+        tar -xvf $BACKUP_DIR/backup.tar -C $BACKUP_VERIFY_PGDATA
+        $OLD_BINARIES/pg_ctl -D $BACKUP_VERIFY_PGDATA -w start
```

Comment:
Is this sufficient? Should there maybe be a `SELECT count(*) FROM some_table` smoke test?

Reply:
And what is the actual purpose of this test? Why are we verifying the **old** version works? Shouldn't we instead verify that the upgraded version works?

Reply:
Starting the server should be sufficient. The need for that is to make sure the backup could be restored, and nothing in between was broken (currently there are very few steps, but still it's a good practice to verify a backup). The reason to verify it with the old version is that if something goes wrong, the backup has to be restored to the old version. Verification that the upgraded version works will be done on the fly by using it.